### PR TITLE
Update Search Endpoint to use HousingSearchAPI

### PR DESCRIPTION
**Pull request by LBHCallumM**
## Link to JIRA ticket

[Ticket](https://www.pivotaltracker.com/story/show/180329889)

## Describe this PR

The URL for search will be changed from `/properties/` to `/properties/search/`. Frontend will need to implement feature toggling to only enable the search to be updated on dev. 

The changes in the AssetAPI will be released by MMH on Thursday, so it will need to be delayed until then. Once released, I can delete the code for the old endpoint. 

Since the old code isn't being removed yet, these changes can be released without breaking anything.

#### _Checklist_

- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

**Comment by teodor-hristoforov-hackney** [comment]
@LBHCallumM you might want to take a look at the CircleCI tests.

File: RepairsApi.Tests/MockApiGateway.cs [hunk 1]
```diff
@@ -194,6 +202,12 @@ public static object HandleAssetAPI()
             return _assetApiResponse;
         }
 
+        [Route("http://testhousingsearchapi/search/assets")]
```
Comment (teodor-hristoforov-hackney) [inline RepairsApi.Tests/MockApiGateway.cs:205]: Should this be pointing to testing?

File: RepairsApi.Tests/V2/Controllers/PropertiesControllerTests.cs [hunk 2]
```diff
@@ -17,49 +17,79 @@
...
+using System;
+using System.Linq;
 
 namespace RepairsApi.Tests.V2.Controllers
 {
     [TestFixture]
     public class PropertiesControllerTests : ControllerTests
     {
-        private Fixture _fixture;
+        private readonly Fixture _fixture = new Fixture();
```
Comment (teodor-hristoforov-hackney) [inline RepairsApi.Tests/V2/Controllers/PropertiesControllerTests.cs:29]: Why set the AutoFixture to readonly?

Comment (LBHCallumM) [reply RepairsApi.Tests/V2/Controllers/PropertiesControllerTests.cs:29]: It doesn't make any difference, it behaves the same as Random. I just thought it looked better as a single line.

File: RepairsApi.Tests/MockApiGateway.cs (hunk 1 above)
Comment (teodor-hristoforov-hackney) [reply RepairsApi.Tests/MockApiGateway.cs:205]: Don't worry about it, it's fine how it is.

File: RepairsApi.Tests/V2/Controllers/PropertiesControllerTests.cs (hunk 2 above)
Comment (teodor-hristoforov-hackney) [reply RepairsApi.Tests/V2/Controllers/PropertiesControllerTests.cs:29]: Tested this and it's working fine, was just overthinking it.

File: RepairsApi.Tests/E2ETests/SearchForPropertiesE2ETests.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        private readonly Random _random = new Random();
+
+        [SetUp]
+        public void SetUp()
+        {
+            MockApiGateway.ForcedCode = null;
+        }
+
+        [Test]
+        public void SearchForPropertiesWhenBadRequestReturnsBadRequest()
```
Comment (Asura5) [inline RepairsApi.Tests/E2ETests/SearchForPropertiesE2ETests.cs:36]: BadRequestReturnsBadRequest I would have thought is a tautology =D
Maybe GatewayReturnsBadRequest, or suchlike?

File: RepairsApi.Tests/E2ETests/SearchForPropertiesE2ETests.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            var apiResponse = new HousingSearchAPIResponse
+            {
+                Assets = _fixture.CreateMany<Asset>(numberOfResults).ToList(),
+                Total = numberOfResults
+            };
+
+            MockApiGateway.SetHousingSearchApiResponse(apiResponse);
+            MockApiGateway.ForcedCode = null;
+
+            ApiGateway client = new ApiGateway(new HttpClientFactoryWrapper(Client));
```
Comment (Asura5) [inline RepairsApi.Tests/E2ETests/SearchForPropertiesE2ETests.cs:97]: This is in all the tests. Refactor available to make it more DRY? (and therefore more understandable)

File: RepairsApi/Startup.cs [hunk 5]
```diff
@@ -259,12 +258,14 @@ public static void RegisterGateways(IServiceCollection services)
...
             services.AddTransient<IAssetGateway, AssetGateway>();
+            services.AddTransient<IHousingSearchGateway, HousingSearchGateway>();
+            services.AddTransient<IPropertyGateway, PropertyGateway>();
         }
 
         public static void RegisterUseCases(IServiceCollection services)
         {
             services.AddTransient<IListAlertsUseCase, ListAlertsUseCase>();
-            services.AddTransient<IListPropertiesUseCase, ListPropertiesUseCase>();
+            services.AddTransient<IHousingSearchUseCase, HousingSearchUseCase>();
```
Comment (Asura5) [inline RepairsApi/Startup.cs:268]: Again, is this feature toggled and would it be wise to do so?

File: RepairsApi/V2/Boundary/Response/SearchForPropertiesResponse.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace RepairsApi.V2.Boundary.Response
+{
+    public class SearchForPropertiesResponse
+    {
+        public long Total { get; set; } = 0;
```
Comment (Asura5) [inline RepairsApi/V2/Boundary/Response/SearchForPropertiesResponse.cs:10]: Is Total necessary? Can't we just count the Properties?

File: RepairsApi/V2/Controllers/PropertiesController.cs [hunk 7]
```diff
@@ -25,48 +25,47 @@ public class PropertiesController : BaseController
...
         [ProducesResponseType(typeof(List<PropertyListItem>), 200)]
         [ProducesResponseType(typeof(string), StatusCodes.Status502BadGateway)]
         [ProducesDefaultResponseType]
         [Authorize(Roles = UserGroups.Agent + "," + UserGroups.ContractManager + "," + UserGroups.AuthorisationManager + "," + UserGroups.Operative)]
-        public async Task<IActionResult> ListProperties([FromQuery] string address, [FromQuery] string postcode, [FromQuery] string q)
+        [Route("search")]
+        public async Task<IActionResult> SearchForProperties([FromQuery] SearchForPropertiesRequest query)
         {
-            PropertySearchModel searchModel = new PropertySearchModel
+            var searchModel = new SearchForPropertiesModel
```
Comment (Asura5) [inline RepairsApi/V2/Controllers/PropertiesController.cs:60]: This looks as though we will have to modify the incoming request from the frontend, is that correct?

File: RepairsApi/V2/Gateways/HousingSearchGateway.cs [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            Uri url = new Uri($"search/assets?{searchModel.GetQueryParameters()}", UriKind.Relative);
+
+            var response = await _apiGateway.ExecuteRequest<HousingSearchAPIResponse>(HttpClientNames.HousingSearch, url);
+
+            if (!response.IsSuccess)
+            {
+                _logger.LogError($"Call to {url} failed with {response.Status}");
+                throw new ApiException(response.Status, Resources.PropertiesFailure);
+            }
+
```
Comment (Asura5) [inline RepairsApi/V2/Gateways/HousingSearchGateway.cs:40]: Might be useful to log a warning if it's a successful search but can't find any matches

File: RepairsApi/Startup.cs (hunk 5 above)
Comment (LBHCallumM) [reply RepairsApi/Startup.cs:268]: When I readded it, I must have moved it to the bottom. It's still in there.

File: RepairsApi/V2/Boundary/Response/SearchForPropertiesResponse.cs (hunk 6 above)
Comment (LBHCallumM) [reply RepairsApi/V2/Boundary/Response/SearchForPropertiesResponse.cs:10]: The Housing Search API uses pagination. Total will be the total number of results, whereas properties will be however many the frontend has requested.

**Review by Asura5 — APPROVED** [review]

**Pull request merged**